lib/Parse/ParseDecl.cpp

```diff
@@ -3926,9 +3927,46 @@ Parser::parseDecl(ParseDeclOptions Flags,
         break;
       }
 
-      const bool IsProbablyFuncDecl =
-          Tok.isIdentifierOrUnderscore() || Tok.isAnyOperator();
+      bool IsProbablyFuncDecl = false;
+
+      if (Tok.isIdentifierOrUnderscore() || Tok.isAnyOperator()) {
+
+        // This is a probe, so backtrack when we're done
+        // and don't emit diagnostics.
+        BacktrackingScope backtrackingScope(*this);
+        DiagnosticSuppression diagnosticSuppression(Diags);
+
+        // TODO make sure we can't get these from the environment
+        SourceLoc staticLoc; 
+        StaticSpellingKind staticSpelling = StaticSpellingKind::None;
+        DeclAttributes attributes;
+        bool hasFuncKeyword = false;
+
+        ParserResult<FuncDecl> parserResult =
+          parseDeclFunc(staticLoc,
+            staticSpelling,
+            Flags,
+            attributes,
+            hasFuncKeyword);
+
+        // it looks like we always get a good parse result.
+        // But if there are no parentheses
+        // the left and right parens will be at the beginning of
+        // the expression and will be equal so we use
+        // this to reject degenerate declarations.
+
+        if (parserResult.isNonNull()) {
+          auto parameterList = parserResult.get()->getParameters();
+          if (parameterList->getLParenLoc() != parameterList->getRParenLoc()) {
+            IsProbablyFuncDecl = true;
+          }
+        }
+        // backtrackingScope and diagnostic suppression are both RAII
+        // when they go out of scope, Tok returns to the beginning of
+        // the identifier, and diagnostics are unsuppressed.
```

Comment:
We generally try to avoid comments which explain _what_ is going on (which can be understood by reading the code) and try to focus on the _why_, unless something is particularly complicated.

I would avoid this comment and similar comments above and below.

```
// This is a probe, so backtrack when we're done
// and don't emit diagnostics.
```

```
// if probe succeeded, then do it for real
```

Reply:
I agree with you that the specific comment you’re marking here (about RAII variables) is unnecessary, but I actually think the comments about probing are helpful “why” comments. (Maybe they could be shortened slightly—“This is a probe, so we will always backtrack after we’re done.” is *definitely* a helpful “why” comment.)